### PR TITLE
Add Custom id

### PR DESCRIPTION
**Pull request by fonzarely**
This make id field configurable. 
I'm not sure on how to test this. Please advise.

This is in response for https://github.com/josx/aor-feathers-client/issues/16

### CODE REVIEW

**Comment by fonzarely** [comment]
Any comment ?

**Comment by josx** [comment]
did you read my review?

copy here agin:
It seems not to work on default behaviour, i think need to return params.id (you are using it for update)
This is about ```return params;```

**Comment by josx** [comment]
Sorry but i replied on review but didnt choose any options, so didnt post it.

File: src/restClient.js [hunk 1]
```diff
@@ -8,23 +8,32 @@ import {
...
   const mapRequest = (type, resource, params) => {
     const service = client.service(resource);
     let query = {};
 
+    const mapParamsId = params => {
+      if(options.id !== 'id'){
+        params[options.id] = params.id;
+        delete params.id;
+      }
+      return params;
```
Comment (fonzarely) [inline src/restClient.js]: Tests are passing, could you help to update tests ?

**Comment by josx** [comment]
why did you add yarn.lock here?
If you want do update your branch (and you are the only devel) you can rebase against master.

File: src/restClient.js (hunk 1 above)
Comment (josx) [inline src/restClient.js]: Tests pass (but here's why):
https://github.com/josx/aor-feathers-client/blob/master/test/restClient.spec.js#L154

For update the test we have to test it with different Ids keys. So we must:

* Alter all json strucutres to support dynamic id key
* Test the usecase for default id key param
* Repeat tests with another id key (fakeId)

Comment (josx) [inline src/restClient.js]: @fonzarely  
You can start viewing this sample code here, it should work (please copy it and if you think its ok use it as base).

https://github.com/josx/aor-feathers-client/blob/fonzarely-custom-id/test/restClient.spec.js

https://github.com/josx/aor-feathers-client/commit/0559705b947dd75605ac82a2f2b897f1e23c4ace

**Comment by fonzarely** [comment]
I guess I failed my rebase. I'll rollback it.

**Comment by 1beb** [comment]
Is this working? Is the only thing left to be done is to add tests?

**Comment by fonzarely** [comment]
I think yes, tests would make sure I did it right. As I'm new to feathers I'm requesting for help on testing.

**Comment by 1beb** [comment]
Why were there so many changes to the yarn file?

**Comment by 1beb** [comment]
ok I see now: 

```
const App = () => (
    <Admin restClient={restClient(feathersClient, { id: '_id'})}>
        <Resource name="reminders" list={RemindersList} create={RemindersCreate} edit={RemindersEdit} remove={Delete} />
    </Admin>
);
```

I'm using the edited version in an application right now and can confirm that it does not work. 

Steps to reproduce:

1. feathers generate app
2. feathers generate service > service > mongoose
3. npm install 
4. npm start
5. start admin-on-rest from test repo

**Comment by 1beb** [comment]
I messed around with this for about 8 hours. I couldn't get it. Sorry, I'm very new to JS, will await PR merge.

**Comment by josx** [comment]
@fonzarely told that works with custom id, it seems he is using it with a custom id. 

I tested with default id and  it seems not working so we realized we need tests.
So i think We need help with tests.

There is a WIP on tests above in the requested changes.

**Comment by 1beb** [comment]
I can confirm that I tested this with a basic mongoose service and it did not work. (with '_id' option set appropriately). Perhaps adding tests that mimic a mongo store would be a good idea. I have no idea how to do that. I think neDB works similarly however.

**Comment by thngdude** [comment]
I believe I tried the same as @1beb and it didn't work. I worked around it by adding hooks to the user.model.js in the feathers app. Not ideal, but works for now.

**Comment by josx** [comment]
@thngdude in meantime you can post here your workaround

**Comment by thngdude** [comment]
Actually, I got a version working that modified restClient.js this morning.
Will share later.

**Comment by thngdude** [comment]
Sorry, my git fu isn't very strong. This is a work in progress and builds on what @josx shared.

```
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _types = require('./lib/rest/types.js');

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = function (client) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : { id: 'id' };

  var mapRequest = function mapRequest(type, resource, params) {
    var service = client.service(resource);
    var query = {};

    var mapParamsId = function mapParamsId(params) {
      if (options.id !== 'id') {
        params[options.id] = params.id;
        delete params.id;
      }
      return params;
    };

    switch (type) {
      case _types.GET_MANY:
        var ids = params.ids || [];
        query[options.id] = { '$in': ids };
        query['$limit'] = ids.length;
        return service.find({ query: query });
      case _types.GET_MANY_REFERENCE:
      case _types.GET_LIST:
        var _ref = params.pagination || {},
            page = _ref.page,
            perPage = _ref.perPage;

        var _ref2 = params.sort || {},
            field = _ref2.field,
            order = _ref2.order;

        var mapField = function mapField(field) {
          return field === 'id' ? options.id : field;
        };
        var sortKey = '$sort[' + mapField(field) + ']';
        var sortVal = order === 'DESC' ? -1 : 1;
        if (perPage && page) {
          query['$limit'] = perPage;
          query['$skip'] = perPage * (page - 1);
        }
        if (order) {
          query[sortKey] = JSON.stringify(sortVal);
        }
        Object.assign(query, params.filter);
        return service.find({ query: query });
      case _types.GET_ONE:
        return service.get(params.id);
      case _types.UPDATE:
        return service.patch(params.data.id, mapParamsId(params.data), params.data);
      case _types.CREATE:
        return service.create(params.data);
      case _types.DELETE:
        return service.remove(params[options.id]);
      default:
        throw new Error('Unsupported FeathersJS restClient action type ' + type);
    }
  };

  var mapResponse = function mapResponse(response, type, resource, params) {
    switch (type) {
      case _types.GET_ONE:
      case _types.UPDATE:
      case _types.DELETE:
        return { data: _extends({}, response, { id: response[options.id] }) };
      case _types.CREATE:
        return { data: _extends({}, params.data, { id: response[options.id] }) };
      case _types.GET_LIST:
        response.data = response.data.map(function (_obj) {
          if (options.id !== 'id')
            _obj.id = _obj[options.id];
          return _obj;
        });
        return response;
      default:
        return response;
    }
  };

  return function (type, resource, params) {
    return mapRequest(type, resource, params).then(function (response) {
      return mapResponse(response, type, resource, params);
    });
  };
};
```

**Comment by josx** [comment]
I am closing this PR, please check https://github.com/josx/aor-feathers-client/pull/26

**Pull request closed without merging**